File: src/api/commitments.ts
import { invoke } from "./gesture";
import type { Verdict } from "@/api/verdict";
import { VERDICT } from "@/api/verdict";
import type { TimeScope, DurationSpec } from "@/api/time-scope";

export interface Commitment {
  id: number;
  title: string;
  parent_type: string;
  parent_id: number;
  verdict: Verdict;
  // Null inherits the nearest scoped ancestor's. Unlike every other kind, the *effective* window
  // may not be absent: a commitment with no window anywhere above it is refused at write time.
  time_scope: TimeScope | null;
  // The Verdict Window: how long past the end of its window this stays answerable, as a count of
  // any scope kind. Absent inherits the nearest ancestor Commitment's; nothing above setting one
  // means it never expires.
  verdict_window?: DurationSpec;
  tag_ids: number[];
  position: number;
  is_private: boolean;
  // The bd issue this commitment is tracked as. Written only by the MCP server.
  beads_id?: string;
}

export interface CreateCommitmentRequest {
  title: string;
  parent_type: string;
  parent_id: number;
  verdict?: Verdict;
  time_scope?: TimeScope;
  verdict_window?: DurationSpec;
}

export interface UpdateCommitmentRequest {
  title?: string;
  // Recording, changing and clearing a verdict are all this one field — a misclick is undone by
  // the same call that made it.
  verdict?: Verdict;
  // Absent = leave unchanged, null = clear, value = set.
  time_scope?: TimeScope | null;
  // Absent = leave unchanged, null = go back to inheriting, value = set.
  verdict_window?: DurationSpec | null;
  parent_type?: string;
  parent_id?: number;
  position?: number;
  is_private?: boolean;
}

export async function listCommitments(): Promise<Commitment[]> {
  return invoke<Commitment[]>("list_commitments");
}

export async function getCommitment(id: number): Promise<Commitment> {
  return invoke<Commitment>("get_commitment", { id });
}

/**
 * Creates a commitment.
 *
 * Refused when it would have no effective Time Scope — neither its own nor a scoped ancestor's.
 * That is the rule unique to this kind: a rule held over no window can never come due.
 */
export async function createCommitment(request: CreateCommitmentRequest): Promise<Commitment> {
  return invoke<Commitment>("create_commitment", { request });
}

export async function updateCommitment(
  id: number,
  request: UpdateCommitmentRequest,
): Promise<Commitment> {
  return invoke<Commitment>("update_commitment", { id, request });
}

export async function deleteCommitment(id: number): Promise<void> {
  return invoke<void>("delete_commitment", { id });
}

export async function addTagToCommitment(commitmentId: number, tagId: number): Promise<void> {
  return invoke<void>("add_tag_to_commitment", { commitmentId, tagId });
}

export async function removeTagFromCommitment(commitmentId: number, tagId: number): Promise<void> {
  return invoke<void>("remove_tag_from_commitment", { commitmentId, tagId });
}

/**
 * The verdict a control should write when it is pressed on a commitment currently reading
 * `current`.
 *
 * The tick and the cross are two explicit, equal choices, and each one toggles: pressing the tick
 * on a kept commitment clears it, pressing the cross on a broken one clears it. What neither
 * control ever does is move straight from one verdict to the other — reaching Broken is always
 * the cross, never a repeat of the tick.
 */
export function verdictAfterPressing(pressed: Exclude<Verdict, "unresolved">, current: Verdict): Verdict {
  return current === pressed ? VERDICT.UNRESOLVED : pressed;
}

/**
 * The verdict Enter advances to (Unresolved → Kept → Broken → Unresolved).
 *
 * One key walks the whole answer, so a commitment can be judged, corrected and un-judged without
 * leaving the row. Kept leads because it is the answer given most often; Broken sitting one press
 * further along is not the only way to reach it, since `X` writes Broken directly from any state.
 */
export const NEXT_VERDICT: Record<Verdict, Verdict> = {
  unresolved: "kept",
  kept: "broken",
  broken: "unresolved",
};

File: src-tauri/src/commands/commitments.rs
//! Tauri commands for commitment operations.
//!
//! Thin, like every other command module: open a session, delegate, commit. Note that there is
//! no `set_verdict` command — recording, changing and clearing a verdict are all one ordinary
//! [`update_commitment`] write, so a misclick is undone by the same call that made it.

use tauri::State;

use crate::{
    database::session::SessionFactory,
    error::WireError,
    nodes::{id::NodeId, write},
    tasks::model::{Commitment, CommitmentId, CreateCommitmentRequest, UpdateCommitmentRequest},
};

/// Creates a new commitment.
///
/// Refused when the commitment would have no effective Time Scope — neither its own nor a scoped
/// ancestor's. That is the one write rule unique to this kind: a rule held over no window has
/// nothing to be kept or broken over.
#[tauri::command]
pub async fn create_commitment(
    factory: State<'_, SessionFactory>,
    request: CreateCommitmentRequest,
) -> Result<Commitment, WireError> {
    let mut db = factory.begin().await.map_err(WireError::from_error)?;
    let commitment = crate::tasks::create_commitment(&mut db, request)
        .await
        .map_err(WireError::from_error)?;
    db.commit().await.map_err(WireError::from_error)?;
    Ok(commitment)
}

/// Fetches a commitment by id.
#[tauri::command]
pub async fn get_commitment(
    factory: State<'_, SessionFactory>,
    id: i64,
) -> Result<Commitment, WireError> {
    let mut db = factory.connect().await.map_err(WireError::from_error)?;
    db.commitments()
        .get(CommitmentId(id))
        .await
        .map_err(WireError::from_error)
}

/// Lists every commitment — the Commitment virtual table: stored rows and every commitment
/// Habit's iterations.
#[tauri::command]
pub async fn list_commitments(
    factory: State<'_, SessionFactory>,
    now: chrono::NaiveDateTime,
) -> Result<Vec<Commitment>, WireError> {
    let mut db = factory.begin().await.map_err(WireError::from_error)?;
    let load = crate::mindmap::load(&mut db, now)
        .await
        .map_err(WireError::from_error)?;
    db.commit().await.map_err(WireError::from_error)?;
    Ok(load.commitments)
}

/// Updates a commitment, stored or derived — including recording, changing or clearing its
/// Verdict.
#[tauri::command]
pub async fn update_commitment(
    factory: State<'_, SessionFactory>,
    id: NodeId,
    request: UpdateCommitmentRequest,
) -> Result<Commitment, WireError> {
    let mut db = factory.begin().await.map_err(WireError::from_error)?;
    let commitment =
        write::update_commitment(&mut db, &id, request, chrono::Local::now().naive_local())
            .await
            .map_err(WireError::from_error)?;
    db.commit().await.map_err(WireError::from_error)?;
    Ok(commitment)
}

/// Deletes a commitment and everything beneath it — or archives a commitment Habit's iteration,
/// which is never deleted.
#[tauri::command]
pub async fn delete_commitment(
    factory: State<'_, SessionFactory>,
    id: NodeId,
) -> Result<(), WireError> {
    let mut db = factory.begin().await.map_err(WireError::from_error)?;
    write::delete(
        &mut db,
        "commitment",
        &id,
        chrono::Local::now().naive_local(),
    )
    .await
    .map_err(WireError::from_error)?;
    db.commit().await.map_err(WireError::from_error)
}

/// Adds a tag to a commitment.
#[tauri::command]
pub async fn add_tag_to_commitment(
    factory: State<'_, SessionFactory>,
    commitment_id: i64,
    tag_id: i64,
) -> Result<(), WireError> {
    let mut db = factory.connect().await.map_err(WireError::from_error)?;
    db.commitments()
        .add_tag(CommitmentId(commitment_id), tag_id)
        .await
        .map_err(WireError::from_error)
}

/// Removes a tag from a commitment.
#[tauri::command]
pub async fn remove_tag_from_commitment(
    factory: State<'_, SessionFactory>,
    commitment_id: i64,
    tag_id: i64,
) -> Result<(), WireError> {
    let mut db = factory.connect().await.map_err(WireError::from_error)?;
    db.commitments()
        .remove_tag(CommitmentId(commitment_id), tag_id)
        .await
        .map_err(WireError::from_error)
}
